### ISSUE TITLE
VisualStudioUnresolvedAnalyzerReference not overriding Display property provokes VS 16.3 crash.

### ISSUE THREAD

**Issue by tamlin-mike** [dotnet/roslyn#39092]
Regression from what shipped with VS 16.2.

https://github.com/dotnet/roslyn/blob/7f60b3849b6dd1c3ee64cd2d327cfc5ea6fa7b64/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioAnalyzer.cs#L135

does not override the `Display`property, meaning whoever uses that property will get a null value back despite the comment in `AnalyzerReference` suggesting the opposite.

Enter VS 16.3's `FxCopAnalyzersSuggestedActionCallback.IsNuGetInstalled()`

https://github.com/dotnet/roslyn/blob/829fe7d4fe75b243fc535454cf5c24c173dfe1ec/src/VisualStudio/Core/Def/Implementation/Diagnostics/FxCopAnalyzersSuggestedActionCallback.cs#L142-L144

Crash. Unhandled .NET exception.

Sure, it's VS acting up (and perhaps especially bad it's not handling this exception), but it seems this code is not doing what it `Should` (to quote)
https://github.com/dotnet/roslyn/blob/7f60b3849b6dd1c3ee64cd2d327cfc5ea6fa7b64/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerReference.cs#L29-L31

A quick (and dirty?) "fix" could be to just add an override for that property, forwarding to `_underlying.Display`, since that seems to have a non-null string, but I also believe this code needs a bit more unit testing, the baseclass comment "should" could benefit from some clarification (is it supposed to be dependable non-null, or is it an optional suggestion), and

@mavasani  that FxCop code needs to be fixed to not blindly trust reference values.

Since this is currently a show-stopper for us (we're at an effective stand still, not able to trust VS to not crash while editing code), perhaps that could be incentive to fast track a fix?

Edit: Just in case it helps understanding the issue, `_underlying.Display` contains the string
`Unresolved: MyProjPath\\mscorlib.dll`, where MyProjPath is the local filesystem absolute (with doubled-up backslashes) path to the directory holding the .csproj, which seems obvious to me since there are no dll's at all in that directory.

**Comment by tamlin-mike** [comment]
While I was able to build a replacement dll, I couldn't get it to load in VS. In fact, using it as a drop-in replacement stopped VS from loading C# projects at all. :/

Anyway, this tiny patch should fix the immediate error, but perhaps there's a design error with this that needs further analysis and fixing.

```
--- a/src/VisualStudio/Core/Def/Implementation/Diagnostics/FxCopAnalyzersSuggestedActionCallback.cs
+++ b/src/VisualStudio/Core/Def/Implementation/Diagnostics/FxCopAnalyzersSuggestedActionCallback.cs
@@ -141,6 +141,12 @@ private bool IsNuGetInstalled()

                 foreach (var analyzerReference in document.Project.AnalyzerReferences)
                 {
+                    // Don't crash if Display is null (baseclass only returns null, and at least
+                    // one subclass don't override that
+                    if (analyzerReference.Display == null)
+                    {
+                        continue;
+                    }
                     if (ChildAnalyzerNuGetPackageIds.Contains(analyzerReference.Display))
                     {
                         // We set installed to ensure we don't go through this again next time a
```

**Comment by mavasani** [comment]
I will have a PR for this later today.

**Comment by tamlin-mike** [comment]
@mavasani Excellent. Is there somewhere to get a release-build of the dll once fix is included, that could be dropped in (and ngen'd) in an existing VS installation, or will the only way to get it be to wait for a new VS point release?

**Comment by mavasani** [comment]
@tamlin-mike I'd suggest waiting for the next 16.4 preview release which has the fix. @jinujoseph @vatsalyaagrawal the current Roslyn master branch targets which 16.4 preview release?

**Comment by mavasani** [comment]
@tamlin-mike You can also generate a release build out of 16.3 branch: https://github.com/dotnet/roslyn/tree/release/dev16.3 by cherry picking this commit: https://github.com/dotnet/roslyn/pull/39097/commits/7dfb0816e3bad7400171179a006b6993576d1410

You probably want to build Roslyn.sln with `/p:Configuration=release /p:DotNetUseShippingVersion=true` to get the correct assembly and file versions.

**Comment by tamlin-mike** [comment]
> You can also generate a release build out of 16.3
@mavasani Thank you! I had been skimming the documentation to see if it was possible, but didn't find anything obvious on the "Build" wiki page. This information seems to be exactly what I needed.

**Comment by tamlin-mike** [comment]
@mavasani Editing global.json to remove `-preview6-012264` from the dotnet version, I got it to compile. Unfortunately, comes time to ngen it fails with "Strong name validation failed". Not sure what _exactly_ is provoking that. Possibly some mismatch in arcade version, since it seems the signing key is in an external nuget with such a name.

Before realizing that, I diffed the more accessible info from the dll's

16.3:
[assembly: AssemblyFileVersion("3.300.119.46102")]
[assembly: AssemblyInformationalVersion("3.3.1-beta3-19461-02+2fd12c210e22f7d6245805c60340f6a34af6875b")]
[assembly: AssemblyVersion("3.3.0.0")]

Home-built, tested with both cmd-line build and from within VS:
[assembly: AssemblyFileVersion("42.42.42.42424")]
[assembly: AssemblyInformationalVersion("3.3.1-dev")]
[assembly: AssemblyVersion("42.42.42.42")]

Funny that 16.3 shipped with beta3-19461-*. I would have expected it to be from release/dev16.3.

I guess the 42's are some black magic override, since in the properties in VS it displays the expected version.

Oh well, it was worth giving it a shot, even if it ultimately failed.

Thanks for the help.

**Comment by tamlin-mike** [comment]
I finally managed to get the home-built loaded into VS by creating a full vsix and installing that (into LOCALAPPDATA). That should teach me trying to modify things like this at a file level.

@jinujoseph As far as I'm concerned, this issue can be closed, unless it's waiting for #39097 to get deployed.

@mavasani Again, thanks for the help.

**Comment by jinujoseph** [comment]
also reported at [DC](https://developercommunity.visualstudio.com/content/problem/761989/vs-crash-during-source-code-editing-due-to-fxcop-b.html)
